### PR TITLE
[SPMD] Add embedding dense backward prop rule for postional embedding

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100038
* #100037

### CODE REVIEW

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 1]
```diff
@@ -74,6 +74,15 @@ def embedding_dense_backward_rules(op_schema: OpSchema) -> OutputSharding:
         return OutputSharding(
             output_spec=DTensorSpec(mesh=indices.mesh, placements=[_Partial()])
         )
+    elif (
+        grad_output.placements[0].is_partial() and indices.placements[0].is_replicate()
+    ):
+        # The embedding table is replicated and the indices is also replicated
```
Comment (mrshenli) [inline torch/distributed/_tensor/ops/embedding_ops.py:80]: hmm, why indices are also replicated?

And this assumes the placements list only contains one element?

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 2]
```diff
@@ -74,6 +74,15 @@ def embedding_dense_backward_rules(op_schema: OpSchema) -> OutputSharding:
...
+        # (local is a more precise term). This is postional embedding. In this
+        # case, gradients for the embmedding table should be Partial.
+        return OutputSharding(
+            output_spec=DTensorSpec(mesh=indices.mesh, placements=[_Partial()])
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:84]: Can we add some tests to test this propagation works as expected?

File: torch/distributed/_tensor/ops/embedding_ops.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/_tensor/ops/embedding_ops.py:80]: The code is like `pos = pos_emb(inp.size()[1])` where `pos_emb` is a nn.Embedding. It is not related to batch dimension so it is replicated (local may be more precise).

File: test/distributed/_spmd/test_tracing.py [hunk 3]
```diff
@@ -839,6 +840,44 @@ def train_step(mod, opt, ids, tgt):
...
+                self.wte = nn.Embedding(N, D)
+                self.wpe = nn.Embedding(Block, D)
+                self.norm = nn.LayerNorm(D, elementwise_affine=False)
+                self.fc = nn.Linear(D, D)
+
+            def forward(self, ids, tgt):
+                _, t = ids.size()
+                wte = self.wte(ids)
+                wpe = self.wpe(
+                    torch.arange(0, t, dtype=torch.long, device=ids.device).unsqueeze(0)
```
Comment (mrshenli) [inline test/distributed/_spmd/test_tracing.py:860]: I see, this is what you mean by local.

**Review by mrshenli — APPROVED** [review]

**Pull request closed without merging**